### PR TITLE
storage service

### PR DESCRIPTION
**Pull request by tusharsehwagg**
## Types of changes

What types of changes does your code introduce to this project?

_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)

- [x] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist before merging

_Put an `x` in the boxes that apply_

- [x ] These changes have been thoroughly tested.

- [ ] Changes need to be immediately published on npm.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ This monorepo contains the following packages:
 - `eslint-plugin-internal`: ESLint Plugin with customized rules as per requirement and preferences of devs in Groww. [![npm version](https://badge.fury.io/js/@groww-tech%2Feslint-plugin-internal.svg)](https://badge.fury.io/js/@groww-tech%2Feslint-plugin-internal)
 - `stylelint-config`: Standard Stylelint config adopted in Groww. Customized as per requirement and preferences of devs in Groww. [![npm version](https://badge.fury.io/js/@groww-tech%2Fstylelint-config.svg)](https://badge.fury.io/js/@groww-tech%2Fstylelint-config)
 - `ui-toolkit`: Standard UI Library adopted in Groww. Customized as per requirement and preferences of devs in Groww. [![npm version](https://badge.fury.io/js/@groww-tech%2Fui-toolkit.svg)](https://badge.fury.io/js/@groww-tech%2Fui-toolkit)
+- `web-storage`: Web Storage is a service used that exposes methods to get full control over storage mechanisms available via bucketing.
```
Comment (vikaz-singh) [inline README.md:32]: Web storage is a service that provides simple APIs to get/set data to Local Storage and Cookies. Support for further browser storage is coming soon.

File: packages/web-storage/.eslintignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# npm
+node_modules
+package-lock.json
+npm-debug.log*
+
```
Comment (vikaz-singh) [inline packages/web-storage/.eslintignore:5]: We will have to update this file to ignore pnpm-lock file as well. See if anything else also needs to be added here.

File: packages/web-storage/src/core.ts [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+ * ```
+ * @internal
+ */
+
+import { clearBucketStorage } from './index';
+import { BUCKETS } from './constants';
+
+
+// Suffix for the key name on the expiration items in localStorage
+const CACHE_SUFFIX = '-exptime';
```
Comment (prashant-adhikari-groww) [inline packages/web-storage/src/core.ts:26]: should we move to constant if this is generic ?

File: packages/web-storage/src/core.ts [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+  localStorage.setItem(key, value);
+}
+
+
+function removeItem(key: string) {
+  localStorage.removeItem(key);
+}
+
+
+function eachKey(fn: any) {
```
Comment (prashant-adhikari-groww) [inline packages/web-storage/src/core.ts:151]: name can be better. Also what this function does. Can add comments

File: packages/web-storage/src/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+ * ```
+ * @internal
+ *
+ */
+
+export const calculateBucketSizeInLocalStorage = (bucket: string) : string | null => {
+  if (typeof (Storage) !== 'undefined') {
+    let data = '';
+
+    for (let index = 0; index < localStorage.length; index++) {
```
Comment (prashant-adhikari-groww) [inline packages/web-storage/src/helpers.ts:115]: Let's discuss this iteration on office. 
Let's see if we can use different data structures

File: packages/web-storage/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+ */
+
+export function getDataFromStorage(key: string, storageType: string, bucket = BUCKETS.OTHERS): string | Object | null {
+  const bucketKey = getFullKeyForItem(key, bucket);
+
+  if (checkForErrors() || !localStorageInstance.supported()) {
+    return null;
+  }
+
+  if (storageType === STORAGE_TYPE.LOCAL_STORAGE) {
```
Comment (prashant-adhikari-groww) [inline packages/web-storage/src/index.ts:33]: switch should work here

**Review by prashant-adhikari-groww — COMMENTED** [review]
Good PR. Few things are there, let's have a short syncup for the same in office

File: packages/web-storage/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+ * This method is a part of our Storage Library.
+ *
+ * @returns void
+ *
+ * @internal
+ *
+ */
+
+function clearStorageCookies() {
+  sessionStorage.clear();
```
Comment (abhishekjhagroww) [inline packages/web-storage/src/index.ts:215]: need to clear cookies here.

File: packages/web-storage/package.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  "publishConfig": {
+    "access": "public"
+  },
+  "bugs": {
+    "url": "https://github.com/Groww/webster/issues"
+  },
+  "homepage": "https://github.com/Groww/webster/tree/develop/packages/web-storage#readme",
+  "devDependencies": {
+    "@groww-tech/eslint-config": "^1.0.4",
+    "@groww-tech/tsconfig": "^1.0.0",
```
Comment (shubhamgavali1998) [inline packages/web-storage/package.json:46]: Use workspace:*

File: packages/web-storage/src/constants.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  SESSION_STORAGE: 'SESSION_STORAGE',
+  COOKIE: 'COOKIE',
+  LOCAL_COOKIE_STORAGE: 'LOCAL_COOKIE_STORAGE'
+};
+
+export const BUCKET_SEPERATOR = '$$%&';
+
+export const BUCKET_UNAVAILABLE_ERROR = 'No Bucket Associated!';
+
+export const DEFAULT_STORAGE_EXPIRY_TIME = 10080; // 7 days
```
Comment (shubhamgavali1998) [inline packages/web-storage/src/constants.ts:18]: Display it as 7*24*60

File: packages/web-storage/src/core.ts [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    if (currentTime() >= expirationTime) {
+      removeItem(key);
+      removeItem(exprKey);
+      return true;
+    }
+  }
+}
+
+
+function warn(message: string, err: any) {
```
Comment (shubhamgavali1998) [inline packages/web-storage/src/core.ts:198]: format

File: packages/web-storage/src/helpers.ts [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+ * ```
+ * console.log(getUserKeyFromStoredKey('PERSISTED$$%&kjsdfgkj')); => Output will be 'kjsdfgkj'
+ * ```
+ * ```
+ * console.log(getUserKeyFromStoredKey('someKeyNotHavingOurSeperator')); => Output will be 'someKeyNotHavingOurSeperator'.
+ * ```
+ * @internal
+ */
+
+export const getUserKeyFromStoredKey = (key: string): string => {
```
Comment (shubhamgavali1998) [inline packages/web-storage/src/helpers.ts:83]: naming can be better

File: packages/web-storage/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+  if (checkForErrors() || !localStorageInstance.supported()) {
+    return '';
+  }
+
+  if (STORAGE_TYPE.COOKIE === storageType) {
+    cookie.set(key, data, { expires: expiresInDay, path: '/', secure: true });
+
+  } else if (STORAGE_TYPE.SESSION_STORAGE === storageType) {
+    sessionStorage.setItem(key, data);
```
Comment (shubhamgavali1998) [inline packages/web-storage/src/index.ts:96]: check data type before adding to session storage

File: packages/web-storage/src/index.ts (hunk 6 above)
Comment (tusharsehwagg) [reply packages/web-storage/src/index.ts:39]: i guess, if else will be more readable and english like. that is why chosed this one

**Pull request merged**